### PR TITLE
Use Type manager for ORM and MongoDB ODM or generic handling for other

### PR DESCRIPTION
**Pull request by TomHAnderson**

### CODE REVIEW

File: src/DoctrineModule/Stdlib/Hydrator/DoctrineObject.php [hunk 1]
```diff
@@ -514,16 +519,41 @@ function ($item) {
...
+        if (class_exists('Doctrine\ODM\MongoDB\DocumentManager')
+            && $this->objectManager instanceof DocumentManager) {
+
+            $type = ODMType::getType($typeOfField);
+            $platform = $this->objectManager->getConnection()->getDatabasePlatform();
+
+            return $type->convertToPHPValue($value, $platform);
+        }
+
+        // Generic handling of types (not flexible with custom types)
```
[the remark this answers is not in the document]
Comment (TomHAnderson) [reply src/DoctrineModule/Stdlib/Hydrator/DoctrineObject.php]: CouchDB has or had an ODM implementation.

**Comment by TomHAnderson** [comment]
Tests for all types seems like a function of the dbal or odm implementation because that is where those type classes exist.  Here I'm only using what those already implement and IMHO this is the correct way to do it and the original, replaced, code was very wrong.

But this more correct implementation can cause incorrect implementations to fail.  See
https://github.com/API-Skeletons/zf-doctrine-module-zend-hydrator/issues/6

**Comment by Erikvv** [comment]
I think if the field type is json_array or sinple_array then for that property you should be able to pass an array to the hydrator. 

And if the field type is object you should be able to provide an object. 

You seem to disagree right?

**Comment by Erikvv** [comment]
> Tests for all types seems like a function of the dbal or odm implementation because that is where those type classes exist.

They are tested with data guaranteed to be coming from a database but now we will use them for data coming from other sources.

I am curious what the maintainers of doctrine/dbal think about thrusting this new responsibility on the Type classes and adjusting them for this purpose. They have often been mentioned as performance-critical so I think they will disagree with this use case. 

I have no doubt about your good intentions but I have my doubts about using the Type classes. It definitely needs more approval.

**Comment by TomHAnderson** [comment]
>> And if the field type is object you should be able to provide an object.
>> You seem to disagree right?

I do not disagree but handling that case should be a function of the Type class: https://blog.tomhanderson.com/2018/09/datetime-with-microseconds-for-mysql-in.html
```php
public function convertToPHPValue($value, AbstractPlatform $platform)
    {
        if ($value === null || $value instanceof DateTimeInterface) {
            return $value;
        }
```

In this case on my custom type I cover the case of the value already existing as a DateTimeInterface and I think that's what the other types should do too.

**Pull request closed without merging**